### PR TITLE
feat!: add notification infrastructure

### PR DESCRIPTION
**Pull request by Wielewout**
This PR adds the notification infrastructure as a replacement for #397.

It introduces loki 3.1 with multi-tenancy. To make multi-tenancy work a custom gateway was added. The gateway included in the bitnami chart didn't allow to inject custom nginx config. The official loki chart on the other hand allows multi-tenancy, but only with a single user as the username is used as the organization id (and it has more issues to integrate it).

Because the gateway is defined outside of the loki chart (and because it allows to use tags to enable/disable also logging and notifications independently), also promtail is included separately for the logs. An additional promtail deployment is added to forward notification data from kafka to loki.

The main thing that brings everything together is the secret generated in the loki-gateway chart with all the credentials. Each microservice gets it's own credentials (or even separate credentials for the different tenants in the case of grafana) to access loki gateway. The gateway then sets the headers correctly and forwards the requests.

In case the password is left empty (the default), a random password is generated. Currently it's left in place with a helm uninstall using the annotation `helm.sh/resource-policy: "keep"`, but this should be removed (or made configurable). There is no use case currently for an external user to connect to it, so it can change each time.

- [x] Depends on #457 for the pdb in `loki-gateway`

### CODE REVIEW

**Comment by Wielewout** [comment]
Currently there are issues with the resource requests. When running the `ct install` test locally on a kind cluster we get a maximum amount of resources exceeding the default github actions runner's 4 CPUs. (There is currently also 16 GB of memory and 14 GB of storage available, but that's fine.)

```
$ kubectl describe node kind-control-plane
Name:               kind-control-plane
Roles:              control-plane
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/os=linux
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=kind-control-plane
                    kubernetes.io/os=linux
                    node-role.kubernetes.io/control-plane=
Annotations:        kubeadm.alpha.kubernetes.io/cri-socket: unix:///run/containerd/containerd.sock
                    node.alpha.kubernetes.io/ttl: 0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
CreationTimestamp:  Thu, 18 Jul 2024 21:22:56 +0200
Taints:             <none>
Unschedulable:      false
Lease:
  HolderIdentity:  kind-control-plane
  AcquireTime:     <unset>
  RenewTime:       Wed, 28 Aug 2024 17:19:38 +0200
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  MemoryPressure   False   Wed, 28 Aug 2024 17:19:36 +0200   Thu, 18 Jul 2024 21:22:53 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 28 Aug 2024 17:19:36 +0200   Thu, 18 Jul 2024 21:22:53 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 28 Aug 2024 17:19:36 +0200   Thu, 18 Jul 2024 21:22:53 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 28 Aug 2024 17:19:36 +0200   Wed, 31 Jul 2024 09:49:08 +0200   KubeletReady                 kubelet is posting ready status
Addresses:
  InternalIP:  172.18.0.2
  Hostname:    kind-control-plane
Capacity:
  cpu:                12
  ephemeral-storage:  486903968Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             32512328Ki
  pods:               110
Allocatable:
  cpu:                12
  ephemeral-storage:  486903968Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             32512328Ki
  pods:               110
System Info:
  Machine ID:                 9db69761c578457e9b06e471f9ef836b
  System UUID:                793b4fc0-1dfd-4c1f-b508-b91dd2040971
  Boot ID:                    6c80f142-cd2f-4ec4-87e4-631f7a4e0e6c
  Kernel Version:             6.5.0-1027-oem
  OS Image:                   Debian GNU/Linux 12 (bookworm)
  Operating System:           linux
  Architecture:               amd64
  Container Runtime Version:  containerd://1.7.13
  Kubelet Version:            v1.29.2
  Kube-Proxy Version:         v1.29.2
PodCIDR:                      10.244.0.0/24
PodCIDRs:                     10.244.0.0/24
ProviderID:                   kind://docker/kind/kind-control-plane
Non-terminated Pods:          (49 in total)
  Namespace                   Name                                                       CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                   ----                                                       ------------  ----------  ---------------  -------------  ---
  default                     drax-kuezb5899e-config-api-6d77796b94-bvd9k                0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-cw-ctrl-6d8c8c8d8f-bdzzc                   0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m4s
  default                     drax-kuezb5899e-dashboard-ddc6749bc-55lls                  0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m5s
  default                     drax-kuezb5899e-du-influxdb-0                              0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-du-metrics-server-745d6448d6-28v62         0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m6s
  default                     drax-kuezb5899e-e2-t-ap-5844b67d5c-pphdf                   0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m6s
  default                     drax-kuezb5899e-e2-t-ap-5844b67d5c-prs98                   0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m6s
  default                     drax-kuezb5899e-e2-t-sctp-6cb5c99d56-g5g2h                 0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-e2-t-sm-kpm-6fdb8bd7bd-x4cfp               0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-e2-t-sm-rc-fb889777c-ngcdt                 0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m6s
  default                     drax-kuezb5899e-golang-nkafka-5g-74dd7b759d-vkvnh          0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m5s
  default                     drax-kuezb5899e-grafana-7669996b9c-jw9j6                   0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m4s
  default                     drax-kuezb5899e-influxdb-0                                 0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-influxdb2-0                                0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-kafka-controller-0                         750m (6%)     1125m (9%)  1280Mi (4%)      1920Mi (6%)    2m6s
  default                     drax-kuezb5899e-kafka-controller-1                         750m (6%)     1125m (9%)  1280Mi (4%)      1920Mi (6%)    2m5s
  default                     drax-kuezb5899e-kafka-controller-2                         750m (6%)     1125m (9%)  1280Mi (4%)      1920Mi (6%)    2m5s
  default                     drax-kuezb5899e-kminion-8d465bc95-v8tsc                    0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-kube-state-metrics-759968bd88-xlbkb        0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-loki-compactor-57bb5dcf88-zhvw4            100m (0%)     150m (1%)   128Mi (0%)       192Mi (0%)     2m4s
  default                     drax-kuezb5899e-loki-distributor-76cbc6d4fd-7glrt          100m (0%)     150m (1%)   128Mi (0%)       192Mi (0%)     2m6s
  default                     drax-kuezb5899e-loki-gateway-66f54b8959-764th              100m (0%)     150m (1%)   128Mi (0%)       192Mi (0%)     2m6s
  default                     drax-kuezb5899e-loki-ingester-0                            250m (2%)     375m (3%)   256Mi (0%)       384Mi (1%)     2m6s
  default                     drax-kuezb5899e-loki-querier-0                             100m (0%)     150m (1%)   128Mi (0%)       192Mi (0%)     2m7s
  default                     drax-kuezb5899e-loki-query-frontend-6c68c964d9-wkdbh       100m (0%)     150m (1%)   128Mi (0%)       192Mi (0%)     2m4s
  default                     drax-kuezb5899e-memcachedchunks-0                          100m (0%)     150m (1%)   128Mi (0%)       192Mi (0%)     2m6s
  default                     drax-kuezb5899e-memcachedfrontend-0                        100m (0%)     150m (1%)   128Mi (0%)       192Mi (0%)     2m7s
  default                     drax-kuezb5899e-nats-0                                     0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m6s
  default                     drax-kuezb5899e-network-state-monitor-5d6bb8995-9sh7x      0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m5s
  default                     drax-kuezb5899e-pm-counters-55fd8d6c99-lqpx9               0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m4s
  default                     drax-kuezb5899e-prometheus-node-exporter-cfkjw             0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-prometheus-pushgateway-6d658b4bbd-hw6fp    0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m5s
  default                     drax-kuezb5899e-prometheus-server-0                        0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-promtail-logs-lmlr4                        0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-promtail-notifications-cfc764f6-bzckt      0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-redis-master-0                             100m (0%)     150m (1%)   128Mi (0%)       192Mi (0%)     2m7s
  default                     drax-kuezb5899e-service-monitor-675cdd9977-v8bcl           0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-service-orchestrator-9d86f656b-4smvn       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m5s
  default                     drax-kuezb5899e-telemetry-collector-b88b97bfd-45hn2        0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  default                     drax-kuezb5899e-vector-6f6857f944-z82cv                    0 (0%)        0 (0%)      0 (0%)           0 (0%)         2m7s
  kube-system                 coredns-76f75df574-5g9mm                                   100m (0%)     0 (0%)      70Mi (0%)        170Mi (0%)     40d
  kube-system                 coredns-76f75df574-v8wdl                                   100m (0%)     0 (0%)      70Mi (0%)        170Mi (0%)     40d
  kube-system                 etcd-kind-control-plane                                    100m (0%)     0 (0%)      100Mi (0%)       0 (0%)         40d
  kube-system                 kindnet-szfpd                                              100m (0%)     100m (0%)   50Mi (0%)        50Mi (0%)      40d
  kube-system                 kube-apiserver-kind-control-plane                          250m (2%)     0 (0%)      0 (0%)           0 (0%)         40d
  kube-system                 kube-controller-manager-kind-control-plane                 200m (1%)     0 (0%)      0 (0%)           0 (0%)         40d
  kube-system                 kube-proxy-r7lkm                                           0 (0%)        0 (0%)      0 (0%)           0 (0%)         40d
  kube-system                 kube-scheduler-kind-control-plane                          100m (0%)     0 (0%)      0 (0%)           0 (0%)         40d
  local-path-storage          local-path-provisioner-7577fdbbfb-n9fgx                    0 (0%)        0 (0%)      0 (0%)           0 (0%)         40d
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests      Limits
  --------           --------      ------
  cpu                4250m (35%)   5050m (42%)
  memory             5410Mi (17%)  8070Mi (25%)
  ephemeral-storage  750Mi (0%)    30Gi (6%)
  hugepages-1Gi      0 (0%)        0 (0%)
  hugepages-2Mi      0 (0%)        0 (0%)
Events:              <none>
```

I think that the best we can do is to not set the resources during the drax tests. This can be done within the `ci` directory.

File: charts/drax/Chart.lock [hunk 1]
```diff
@@ -41,9 +41,18 @@ dependencies:
...
+- name: grafana-loki
+  repository: https://charts.bitnami.com/bitnami/
+  version: 4.6.7
+- name: loki-gateway
+  repository: https://accelleran.github.io/helm-charts/
+  version: 0.1.0
+- name: promtail
+  repository: https://accelleran.github.io/helm-charts/
+  version: 6.17.0
+- name: promtail
```
Comment (dimgogos) [inline charts/drax/Chart.lock:53]: I assume the duplication is by accident here

File: charts/drax/values.yaml [hunk 2]
```diff
@@ -386,6 +390,36 @@ pm-counters:
...
+
+  kafkaInitImage:
+    repository: accelleran/acc-generic-img
+    pullPolicy: IfNotPresent
+    tag: "0.8.0"
+
+  image:
+    repository: python
+    pullPolicy: IfNotPresent
+    tag: "3.12.5"
```
Comment (dimgogos) [inline charts/drax/values.yaml]: I guess we could use python:3.12.5-slim here as a default. We can easily test this once deployed E2E.

File: charts/drax/templates/job-drax-notify.yaml
Comment (dimgogos) [inline charts/drax/templates/job-drax-notify.yaml:1]: As briefly discussed, it is not clear why this is needed. If the topic needs to be there, this should be configured on the kafka chart. Not sure if promtail expects also some data to be there. We could verify the behavior in a live setup

**Review by dimgogos — APPROVED** [review]

File: charts/drax/templates/job-drax-notify.yaml
Comment (Wielewout) [reply charts/drax/templates/job-drax-notify.yaml:1]: Only the topic is needed and added with the provisioning config.

File: charts/drax/values.yaml (hunk 2 above)
Comment (Wielewout) [reply charts/drax/values.yaml]: Fixed, but can actually be removed now (see other comment)

**Review by dimgogos — APPROVED** [review]

**Pull request merged**